proposals/p2240.md

```diff
@@ -0,0 +1,689 @@
+element is expected, or directly inside parentheses, and in the latter case, the
+parentheses form a tuple, even if the pack argument has only one element.
+
+A pack expansion block has the form `...{` _statements_ `}`, and is rewritten to
+N copies of the pack template `{` statements `}`, with pack arguments
+substituted as above. `...{` is a single token, and may not contain internal
+whitespace.
+
+Pack expansions currently cannot be nested.
+
+The spelling of `[:]` is taken from the C++ proposal
+[P1858R2](http://wg21.link/P1858R2), but I'm very open to alternatives.
+
+### Examples
+
+```carbon
+// Computes the sum of its arguments, which are i64s
+fn SumInts(..., [:]params: [i64;]) -> i64 {
+  var sum: i64 = 0;
+  for (var i: i64 in params) {
+    sum += i;
+  }
+  return sum;
+}
+```
+
+```carbon
+// Concatenates its arguments, which are all convertible to String
+fn StrCat[T:! [ConvertibleToString;]](..., [:]params: T) -> String {
+  var len: i64 = 0;
+  ...{ len += ([:]params).Length(); }
+  var result: String = "";
+  result.Reserve(len);
+  ...{ result.Append(([:]params).ToString()); }
+  return result;
+}
+```
+
+```carbon
+// Returns the minimum of its arguments, which must all have the same type T
+fn Min[template N:! BigInt if N > 0, T:! Comparable & Value]
+    (..., [:]params: [T; N]) -> T {
+  // Safe since params non-empty
+  var result: T = params[0];
+  for (var x: T in params.Slice(1..)) {
+    if (x < result) {
+      result = x;
+    }
+  }
+  return result;
+}
+```
+
+```carbon
+// Invokes f, with the tuple `args` as its arguments.
+fn Apply[T:! [Type;], F:! CallableWith(..., [:]T)](f: F, args: (..., [:]T)) -> auto {
+  return f(..., [:]args);
+}
+```
+
+```carbon
+fn Zip[ElementTypes:! [Type;]]
+      (..., vectors: Vector([:]ElementTypes))
+      -> Vector((..., [:]ElementTypes)) {
+  var iters: auto = (..., vectors.Begin());
+  var result: Vector((..., [:]ElementTypes));
+  while (...and [:]iters != vectors.End()) {
+    result.push_back((..., *[:]iters));
+    ...{ ([:]iters)++; }
+  }
+  return result;
+}
+```
+
+```carbon
+// Toy example of mixing packs and values in a single parameter list.
+// Takes an i64, any number of f64s, and then another i64.
+fn MiddleVariadic(first: i64, ..., [:]middles: [f64;], last: i64);
+```
+
+## Details
+
+FIXME throughout the following, clarify what parts apply to patterns.
+
+### Statically sized arrays
+
+`[T; N]` is the type of a statically-sized array of `N` objects of type `T`, and
+`[T;]` is a pattern that matches the type `[T; N]` for any `N`. In other words,
+it's a shorthand for `[T; _:! BigInt]`.
+
+FIXME Or maybe we should treat it as an interface/constraint? See
+https://discord.com/channels/655572317891461132/969001583088123905/973681495891910716
+Or maybe it's `[T; template _:! BigInt]` so that we can validly refactor it to
+`[template N:! BigInt]([T; N])`? `auto` seems like it might be ambivalent in a
+similar way, between `_:! Type` and `template _:! Type`.
```

Comment:
#2153 is resolved now. Do I understand correctly that the resolution implies that `auto` does need to be special in this way?

Reply:
`auto` does need to be special so that it matches the template/checked generic attribute of the argument. However, the resolution of #2153 is that both `_:! ...` and `template _:! ...` can match both template and generic arguments. So the choice is dependent on what you want to do with the size value, not what kind of value the caller has. In this case, what sort of type checking needs to be done with the size? The checked generic `_:! BigInt` is likely a better default unless the user specifically asks for a template.